### ISSUE TITLE
fixelcfestats: Catch track file usage (& discussion RE: git tags usage)

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#1834]
If a user tries to run `fixelcfestats` by providing a track file rather than a fixel-fixel connectivity matrix, a more informative error message should be generated explaining the change and directing them to the online documentation.

**Comment by boegel** [comment]
@Lestropie Is the documentation that explains the use of the connectivity matrix rather than a track file already documented? If so, where?

I'm helping someone out who's running into this exact problem using MRtrix 3.0_RC4, it seems:

```
fixelcfestats: [ERROR] required input "tracks_2_million_sift.tck" is not a directory
```

**Comment by Lestropie** [comment]
The online documentation is generated and made available for a number of different software versions, which can be navigated via the green text down in the bottom-left corner of the sidebar. The relevant contents are on the `dev` branch starting [here](https://mrtrix.readthedocs.io/en/dev/fixel_based_analysis/mt_fibre_density_cross-section.html#generate-fixel-fixel-connectivity-matrix).

`3.0_RC4` isn't an announced release; I presume it's there for testing GitHub Actions / binary executable distributions / other "release" stuff. There currently isn't docs generated for that tag, probably because we weren't expecting anybody to be using it.

**Comment by civier** [comment]
Hi Rob,

A quick question about 3.0_RC4:
You say that you did not expect people to use it, but isn't it that by default for new installations, the latest state of master will be installed?

If it is indeed the latest version of master that is being installed, why won't the documentation be updated continously as well to reflect it? is it too heavy on resources?



On Thu, Jan 9, 2020 at 4:44 PM Robert Smith <notifications@github.com<mailto:notifications@github.com>> wrote:

The online documentation is generated and made available for a number of different software versions, which can be navigated via the green text down in the bottom-left corner of the sidebar. The relevant contents are on the dev branch starting here<https://mrtrix.readthedocs.io/en/dev/fixel_based_analysis/mt_fibre_density_cross-section.html#generate-fixel-fixel-connectivity-matrix>.

3.0_RC4 isn't an announced release; I presume it's there for testing GitHub Actions / binary executable distributions / other "release" stuff. There currently isn't docs generated for that tag, probably because we weren't expecting anybody to be using it.

**Comment by Lestropie** [comment]
The latest state of `master` and `3.0_RC4` are two very different software versions. `3.0_RC4` was tagged based on the state of `dev` at the time; as such, the current `master` is *not* a derivative of `3.0_RC4`. Any up-to-date checkout of `master` should report its version as being derived from `3.0_RC3_latest`, which is another tag being used for release testing.

We wouldn't expect users to *directly check out* the tags `3.0_RC3_latest` or `3.0_RC4`, since we've never alerted users to their existence.

Building the documentation is really fast. But I don't think there's any need to have an explicit docs build for `3.0_RC4` since, as I say, nobody should be checking that out directly. The documentation built from `dev` should be applicable to anyone whose software *reports* being `3.0_RC4` (since that's the most recent tag used to generate the version string for anyone who has checked out `dev` recently).

**Comment by boegel** [comment]
@Lestropie Thank you for clarifying. The fact that `3.0_RC3(_*)` and `3.0_RC4` are two very different versions is quite confusing.

It's easy to assume that the `3.0_RC4` listed at https://github.com/MRtrix3/mrtrix3/releases just wasn't tagged as a release, using a sufficiently different tag would have made more sense I think.

So, with all that in mind, which tag/branch is considered to be the latest stable version, which people should be using right now? Is that `3.0_RC3_latest`, or `master` (and if it's the latter, how different is `master` from `3.0_RC3_latest`)?

**Comment by Lestropie** [comment]
Yes, more developer-oriented tag names may have been preferable. Wasn't me :man_shrugging: 

Latest version of `master` is always the latest (& most) stable version, as it includes all latest hotfixes. Checking out `master` and building right now will report as "`3.0_RC3_latest-68-gd411e6c5`": the *most recent* tag is "`3.0_RC3_latest`", but there are 68 commits subsequent to that, which will all be bug fixes (we don't merge any non-egregious-bug behaviour-altering changes to `master`). If you were to check out tag `3.0_RC3_latest` *directly*, you would be missing those 68 bugfix commits. If you were to check out `3.0_RC3` directly, you'd be missing the last 284 bug-fix commits on `master`.

After release we won't have this divergence between `master` and the most recent tag. `master` will point to the most recent release tag. Any hotfixes, instead of going to `master` without any corresponding tag, will instead be tagged according to semantic versioning and generate a new release.

**Comment by jdtournier** [comment]
Sorry, that one was my fault... I hadn't envisaged that people would check out these tags rather than `master` directly. I agree it would have been preferable to use a different tag name for this. I might remove that tag altogether if it creates problems. Most of the required testing for building of binaries and CI is done, there's no particular reason to hold on to the tag, especially as it seems to generate confusion... 

But as @Lestropie says, this is going to be resolved fairly soon, hopefully. We're hoping to release a full-blown `3.0` in the coming weeks.

**Comment by boegel** [comment]
@jdtournier Please don't remove the tag, that's going to create even more confusion...

**Comment by civier** [comment]
Hi Rob and Donald,

I say rename the tag if possible, clarifying it points to dev -- I personally know of an HPC site that installed it recently.

Other than that, now things are MUCH clearer. I think it will be super-useful to users if this information wil be available on the main documentation (like Kenneth, I assumed that all tags are on master).
Also, can you also clarify to me what was the purpose of 3.0_RC3_latest? If you always recommend to check out the latest version of master, so what is special about the point in time when 3.0_RC3_latest was tagged? and why did you generate documentation for it? (if you did not announce it)

Another thing that would be useful is to have an easy way to go from the release in github to the release announcement (or alternatively, to have the version announcement as part of the docs). Sorry if these things already exist, but even if they are, it is not obvious how to find them.


On Thu, Jan 9, 2020 at 8:22 PM Kenneth Hoste <notifications@github.com<mailto:notifications@github.com>> wrote:

@jdtournier<https://github.com/jdtournier> Please don't remove the tag, that's going to create even more confusion...

**Comment by boegel** [comment]
@civier Renaming the tag is just as bad as removing it, no need to make things more confusing than they are already? :)

**Comment by jdtournier** [comment]
> @jdtournier Please don't remove the tag, that's going to create even more confusion...

Yes, I was wondering if that might be the case. I'm not sure what else to do here though, I clearly wasn't expecting users to install it like this. Naive of me, I guess... The simplest thing here would be a rename, as @civier suggests, but it's not simple since there needs to be a (very minor) code change to match. So I can't simply re-tag the existing version.

What I've done is to edit the release on GitHub with hopefully sufficient warnings... [See what you think](https://github.com/MRtrix3/mrtrix3/releases).

**Comment by jdtournier** [comment]
Just to answer a few more questions (and eat hefty amounts of humble pie):

> I think it will be super-useful to users if this information wil be available on the main documentation (like Kenneth, I assumed that all tags are on master).

This is temporary glitch, I wasn't intending for this to happen. I'll make sure it doesn't happen again... If I need to tag versions not intended for release, I'll make sure they're not labelled with tags that look official. 

> Also, can you also clarify to me what was the purpose of 3.0_RC3_latest?

It's a similar issue, really. I wanted to distribute pre-compiled, but relatively up to date versions of the software for a workshop, and a release seemed like the right place to add the binaries. It's really just the latest version of `master` at that point. Again, I don't foresee needing this in future once `3.0` is out, and if I do, I'll make sure to do this on my own fork...

> and why did you generate documentation for it? (if you did not announce it)

I'm not sure we did? The `latest` version of the docs is that generated from the tip of the `master` branch, and gets updated with every push to `master`. There's no `3.0_RC3_latest` version on there as far as I can see? 

> Another thing that would be useful is to have an easy way to go from the release in github to the release announcement (or alternatively, to have the version announcement as part of the docs). Sorry if these things already exist, but even if they are, it is not obvious how to find them.

That's a good suggestion. I'll add a link now for those that have an announcement...

--- 

And just in case it wasn't clear from the above: please accept my sincere apologies for the confusion. This is clearly pretty amateurish on my part...

**Comment by civier** [comment]
First of all, apology accepted (but you do owe me a favor, as my collaborator invested quite a lot of time in an analysis that we will probably have to repeat with a good version of MRtrix :-( ).

Also, if you want to be on the safe side, let people know of the glitch in an announcement on the blog or a broadcast message on the forum. Maybe there are other groups in the same situation.

My replies are below:


Just to answer a few more questions (and eat hefty amounts of humble pie):

I think it will be super-useful to users if this information wil be available on the main documentation (like Kenneth, I assumed that all tags are on master).

This is temporary glitch, I wasn't intending for this to happen. I'll make sure it doesn't happen again... If I need to tag versions not intended for release, I'll make sure they're not labelled with tags that look official.

I actually meant that the information in this thread as a whole should be summarised in the docs:

1. Clarifying that checking out MRtrix gives you the current state of master (i.e., latest bug fixes). Some people (me included) will evade installing a fresh version, specifically if it goes to a cluster. I better have a version with documented bugs than a version where there was not even a chance for people to reveal them. As the saying goes (I just made it now): every bug fix introduces a bug to fix.

2. Clarifying the purpose of release candidates (introducing new features).

3. Clarifying that only releases with announcements should be installed (if one chooses not to install the tip of master)

And so on ...
(If you like these ideas, can I open an issue on GitHub to have a contribution?)

Also, can you also clarify to me what was the purpose of 3.0_RC3_latest?

It's a similar issue, really. I wanted to distribute pre-compiled, but relatively up to date versions of the software for a workshop, and a release seemed like the right place to add the binaries. It's really just the latest version of master at that point. Again, I don't foresee needing this in future once 3.0 is out, and if I do, I'll make sure to do this on my own fork...

Got it. Can you just change the title of the release in GitHub to MRtrix 3.0 rather than MRtrix3? Otherwise, it is a bit confusing.

and why did you generate documentation for it? (if you did not announce it)

I'm not sure we did? The latest version of the docs is that generated from the tip of the master branch, and gets updated with every push to master. There's no 3.0_RC3_latest version on there as far as I can see?

"latest" is really a confusing label, considering the "3.0_RC3_latest" label. I think it should be clarified somewhere that "latest" refer to the tip of master rather than the latest release. Actually, adding to "latest" the date of the master tip will be the best, if that possible.

Another thing that would be useful is to have an easy way to go from the release in github to the release announcement (or alternatively, to have the version announcement as part of the docs). Sorry if these things already exist, but even if they are, it is not obvious how to find them.

That's a good suggestion. I'll add a link now for those that have an announcement...

________________________________

And just in case it wasn't clear from the above: please accept my sincere apologies for the confusion. This is clearly pretty amateurish on my part...

**Comment by boegel** [comment]
> What I've done is to edit the release on GitHub with hopefully sufficient warnings... [See what you think](https://github.com/MRtrix3/mrtrix3/releases).

That makes it pretty clear, without doing any damage (in terms of removing the tag), so :+1: imho, best possible remedy.

**Comment by jdtournier** [comment]
> First of all, apology accepted (but you do owe me a favor, as my collaborator invested quite a lot of time in an analysis that we will probably have to repeat with a good version of MRtrix :-( ).

Yes, I can appreciate this isn't great.. We've only merged this tag 5 weeks ago though, hopefully they haven't invested too much effort yet...? 

More to the point, there's probably no reason to repeat the analysis: there's nothing 'bad' about this version - no overt bugs that would impact the validity of the analysis (at least as far as know, depending on what the issue is [this post](https://community.mrtrix.org/t/dwifslpreproc-eddy-mask-nii-looks-faulty/3164?u=jdtournier)...). I'd probably hold fire on re-doing the entire thing until you've got a better idea of whether there's any need for that. 

> "latest" is really a confusing label, considering the "3.0_RC3_latest" label. I think it should be clarified somewhere that "latest" refer to the tip of master rather than the latest release. Actually, adding to "latest" the date of the master tip will be the best, if that possible.

This is the label given by readthedocs by default. We can probably change it to e.g. `master`, but that will probably break all kinds on links on the forum. And it has to remain static, since it's a permalink... So I vote it stays as it is. 

But I do agree it's annoying that the docs themselves don't have a more explicit reference to the exact version they refer to, and the date they were generated. Maybe there's a way to do that with readthedocs' restructured text markup, we'd need to investigate.

**Comment by thijsdhollander** [comment]
> Also, if you want to be on the safe side, let people know of the glitch in an announcement on the blog or a broadcast message on the forum. Maybe there are other groups in the same situation.

@civier yes, indeed.  I know of a couple of places where cases of "RC4" popped up, but I've just managed them myself to avoid problems and lost time personally.  I don't know what the intended strategy was/is, so I'm not touching the subject beyond that.

**Comment by civier** [comment]
Something that still strikes me is why github does not print out the branch of the tag?
I can open an issue on github, but since you know github much better than I do -- do you think that it is intentional that the branch is not listed next to the tag?


On Fri, Jan 10, 2020 at 10:55 AM Thijs Dhollander <notifications@github.com<mailto:notifications@github.com>> wrote:

Also, if you want to be on the safe side, let people know of the glitch in an announcement on the blog or a broadcast message on the forum. Maybe there are other groups in the same situation.

@civier<https://github.com/civier> yes, indeed. I know of a couple of places where cases of "RC4" popped up, but I've just managed them myself to avoid problems and lost time personally. I don't know what the intended strategy was/is, so I'm not touching the subject beyond that.

**Comment by thijsdhollander** [comment]
> Something that still strikes me is why github does not print out the branch of the tag?

No, that makes sense though: there's actually no concept of "the (unique) branch of the tag", so as to say.  A tag points to a single commit, but a single commit can be "part" of many branches.  The closest you can get to seeing the kind of information you're wondering about is this:

- From the release/tags page, click on the commit that the tag points to.  You can see this on the left, right below the name of the tag.  In case of RC4, that takes you to this commit: https://github.com/MRtrix3/mrtrix3/commit/b774c82e2df94fa3d76480389f205e78226c6dca

- On that page you can see which branches "have" this commit somewhere in their history of commits.  Depending on what kind of development or branching model developers use, this information can be more or less well structured or helpful.

**Comment by jdtournier** [comment]
OK, I've made the announcement - I hadn't realised this was already affecting that many people. Apologies once again. 

Regarding the rationale behind this tag: there is none, it was a dumb mistake, and ridiculously thoughtless on my part. For the record, this tag started life on my own fork, where I was testing the use of GitHub Actions to automate the creation of binary installers for the various platforms we target. I needed a fresh tag to trigger the builds, and that's the one I used on my fork. Once it was all up and running, I created a [pull request](https://github.com/MRtrix3/mrtrix3/pull/1824) for it and it was merged, tag and all. It hadn't occurred to me by then that it would obviously look like the next 'official' version to any external user... I'll fully admit that was a pretty dumb thing to do. All I can do is apologise and endeavour not to screw up like this again.

**Comment by civier** [comment]
Thanks Donald.

To be constructive, it is clear that many groups are installing the released (or tagged) versions rather than the tip of master as you recommend.
I am not sure why it is so (maybe this is how things are being done in most other open source?), but if it is not your intention, you should educate people (apparently the install instructions are not enough).
I suggest to add documentation that explains the logic behind the release cycles, with the hope that people will understand why it is beneficial for them to install the latest commit to master:
https://github.com/MRtrix3/mrtrix3/issues/1871

As I explain in the above issue, my reason for not installing the tip of master is the potential of being exposed to undocumented/undiscovered bugs due to recent bug fixes. I would be happy therefore if the release cycle / install recommendations will be modified a bit, and I wonder if you'd be interested to open it for discussion? That said, I understood from Rob that the release scheduling will be altered after the official release of MRtrix 3.0 anyway, so maybe there is no point. Your call.

**Comment by jdtournier** [comment]
No, I completely agree that we should encourage users to stick to tagged versions - this is how it's supposed to be done. What is true is that because we've always recommended users stick to the `master` branch, I haven't so far given enough consideration to the obvious reality that users will (rightly) be going for tags - this might have contributed to this debacle. But it should absolutely be the case that official-looking tags should be considered stable. This is 100% my goof, and goes against our own _de facto_ policy. 

I'd personally really like to remove this tag given that it's causing more widespread confusion than I'd realised. But I understand the rationale for keeping it there now that it's been pushed out...

> I suggest to add documentation that explains the logic behind the release cycles, with the hope that people will understand why it is beneficial for them to install the latest commit to master:
#1871

I agree we probably need to do this. The point of tagged releases is as you understand it - no debate there. Once we start releasing precompiled packages, this will be even more important (we can't generate packages for every single push to `master` - although we might consider nightly releases, but it's apparently [not trivial on GitHub](https://github.community/t5/How-to-use-Git-and-GitHub/Nightly-Builds-Latest-Branch-Releases/td-p/4922)). 

The reason we would advocate sticking to `master` (at least between tag changes) is that we are careful to only push overt bug fixes and non-behaviour altering changes to `master`. As long as the base tag remains the same (as reported by the `-version` option), then it's essentially the same code with annoying bugs fixed. Any new features or behaviour-altering changes will be pushed to the `dev` branch for the next release.  

> my reason for not installing the tip of master is the potential of being exposed to undocumented/undiscovered bugs due to recent bug fixes.

So we might slightly disagree on that, as per my above comment. I think you're better off sticking to `master` if you want to minimise exposure to bugs. But what is absolutely true is that we can't guarantee that behaviour will remain absolutely unchanged if your pipeline was subject to a bug that was fixed - we're careful to only push bug fixes to `master` if they're necessary and non-behaviour-altering, but there's no getting around the fact that it's different code...

One of the problems we face is that oddly enough, the tip of `master` is likely to be the most bug-free version of the corresponding version - it's had all the bug fixes, and _only_ bug fixes. This is likely to be less of a problem once we start rolling out proper releases (which I'm hoping is imminent), since we'll be in a better position to release bug fix releases more frequently (with micro version increments, e.g. `3.0.1` -> `3.0.2`, as per semantic versioning).

> I would be happy therefore if the release cycle / install recommendations will be modified a bit, and I wonder if you'd be interested to open it for discussion? 

Definitely. Happy to discuss what you have in mind on #1871. Even better if you want to draft something and create a pull request for it. :+1:

This might also be a good candidate for the wiki section of the forum (which I haven't opened up publicly yet, but we will once we release), or the GitHub wiki itself, since it's likely to evolve somewhat over time and shouldn't be release-specific. Otherwise it's painful to make minor changes to the documentation given that each pull request has to pass through automated testing and a round of reviews... Note that it's not either/or: I think we should document the logic/policy for the versioning on the wiki, but also flag these up on the install instructions, with links to those sections of the wiki.

**Comment by Lestropie** [comment]
Closed by #1869. Any subsequent discussion or changes RE: `git` tag usage on this project should be posted as a new Issue, linking to this thread.
